experimental/juvix/contract.jvx

```diff
@@ -0,0 +1,198 @@
+Map = Data.Map
+
+-- common module interfaces for data structures in Common?
+MonoidMeasure : Common.Measure
+MonoidMeasure =
+  type T = Nat
+
+  -- (<>) is normal semi-group  type class function
+  (<>)    = (<>)
+  measure = id
+
+
+-- need OrdMap, as modules that one goes to can't be modules?
+Accounts = Map.OrdMap Data.String MonoidMeasure
+
+-- we send in the string module, as it has a total order
+-- function named compare, that is also part of a general type class
+Token =
+  type Address  = s : String.T {String.length s == 36}
+
+  type Storage = {
+    total-supply : Nat.T;
+    accounts     : Accounts.T { Accounts.measure-value == total-supply }
+  }
+
+  empty-storage : Storage
+  empty-storage = {
+    total-supply = 0;
+    accounts     = Accounts.empty;
+  }
+
+  type T = {
+    storage = Storage;
+    version = Nat.T;
+    name    = String.T;
+    symbol  = Char.T;
+    owner   = Address;
+  }
+
+
+Transaction =
+  type Transfer = {
+    from-account : Token.Address;
+    to-account   : Token.Address;
+    ammount      : Nat.T;
+  }
+
+  type Mint = {
+    mint-amount     = Nat.T;
+    mint-to-account = Token.Address;
+  }
+
+  type Burn = {
+    burn-amount       = Nat.T;
+    burn-from-account = Token.Address;
+  }
+
+  type Data =
+    | Transfer : Transfer -> Data
+    | Mint     : Mint     -> Data
+    | Burn     : Burn     -> Data
+
+  type T = {
+    data               : Data;
+    authorized-account : Token.Address;
+  }
+
+
+has-n : Accounts.T -> Token.Address -> Nat -> Bool
+has-n accounts add to-transfer =
+  case Accounts.select accounts add of
+  | Just n  -> to-transfer <= n
+  | Nothing -> False
+
+
+account-sub : acc : Accounts.T
+           -> add : Token.Address
+           -> num : Nat.T {has-n acc add num}
+           -> Accounts.T
+account-sub accounts add number =
+  case Accounts.select accounts add of
+  | Just balance ->
+     Accounts.put accounts ~key:add ~val:(balance - number)
+
+account-add : Accounts.T -> Token.Address -> Nat.T -> Accounts.T
+account-add accounts add number =
+  Accounts.update accounts ~f:(number +) ~key:add
+
+
+
+transfer-stor : stor   : Token.Storage
+              -> ~from : Token.Address
+              -> ~to   : Token.Address
+              -> num   : Nat.T {has-n stor.accounts from num}
+              -> Token.Storage
+transfer-stor stor ~from:add_from ~to:add_to num =
+  let new-acc = account-add (account-sub stor.accounts add_from) add-to num in
+  -- This is valid, because the monoid and measure of account track this
+  { total-supply = stor.total-supply
+  ; accounts     = new-acc
+  }
+
+
+
+Validation : Module
+  type T = Token.T -> Transaction.T -> Bool
+
+  transfer : T
+  burn     : T
+  mint     : T
+Validation =
+  type T = Token.T -> Transaction.T -> Bool
+
+  mint token tx =
+    case tx.data of
+    | Transaction.Mint ->
+      token.owner == tx-tx-authorized-account
+    | Transaction.Burn _ | Transaction.Transfer _ ->
```

Comment:
Is there a reason this match can't just be `_` ?

Reply:
We want exhaustion. Imagine if we updated the variant wtih another operation, maybe it is a kind of a subset of a mint. It's always important to be exhaustive and  try to never use _

Reply:
OK, sure.